Improvements to esp32 build configuration

UncleGrumpy opened the pull request:
These changes include several enhancements to the ESP32 build configuration and deployment of custom and test builds. The language flavor is determined by the offset of the main.avm partition in the partition table, simplifying Elixir supported builds and making the configuration of the image less error prone. The configuration of an Elixir build may be done without altering git tracked files by using `idf.py -DAVM_ELIXIR_SUPPORTT=on set-target ${CHIP}`. This is the equivalent of setting the custom partition file name to `partitions-elixir.csv` using `idf.py menuconfig`. Other existing AtomVM configuration options have been added to main/Kconfig.projbuild and are exposed in `idf.py menuconfig` under the AtomVM menu to make the configuration process more natural for users already familiar with ESP-IDF and its build and configuration system.

The auto discovered flavor of the esp32boot (`boot.avm` partition) libraries are now included in the `idf.py flash` task. This makes use of the build/mkimage.sh script only necessary for building release images in the CI or to create a complete image for future use, and makes the experience like most other ESP-IDF based projects. To flash only the factory (AtomVM) partition `idf.py app-flash` may be used or `idf.py bootloader-flash app-flash` if changes made affect the bootloader partition as well.

Changes the build configuration for reproducible builds, this strips leading path information (which is replaced by placeholders), and allows locally building from a release tag and generating a binary with the same sha256 checksum as official release binaries (as long as ESP-IDF and toolchain versions match those used to build the release). See:
https://docs.espressif.com/projects/esp-idf/en/v5.5/esp32/api-guides/reproducible-builds.html

The `nvs` partition may now be generated and included in the build/mkimage.sh script (or flashed manually using the `idf.py nvs-flash` command) if the file `nvs_partition.csv` is present in the top level esp32 platform directory. An example is provided and `nvs_partition.csv` is added to `.gitignore` to prevent accidental pushes of sensitive passwords.

Closes atomvm/exatomvm#34

These changes are made under both the "Apache 2.0" and the "GNU Lesser General
Public License 2.1 or later" license terms (dual license).

SPDX-License-Identifier: Apache-2.0 OR LGPL-2.1-or-later

UncleGrumpy committed:
Allow setting ESP32 build options from menuconfig

Exposes the settable build options for the ESP32 platform to the `idf.py menuconfig` utility,
allowing for a more natural configuration experience for those already accustomed to the ESP-IDF
development environment.

Signed-off-by: Winford <winford@object.stream>

petermm commented:
Not sure I like this bundling of commands:

1. Imho it's a coupling, implicit behaviour change - which might hinder future things, like point 2..

2. In a simulator flow I use: `idf.py build && touch ~/path_to_project/atomvm_wokwi_bins/flasher_args.json` (as to build and then reboot sim) - and having mkimage in there would slow things down.

But I'm also a poweruser of `idf.py build && ./build/mkimage.sh` - so I get the drift, but not sure I understand how it simplifies deploying custom images..

bettio commented:
I'm not sure if it is an addition that fills some kind of huge gap in v0.6 (or not), so regardless to the outcome of this discussion I would target main branch anyway.

bettio reviewed (commented):
I'm not sure if it is an addition that fills some kind of huge gap in v0.6 (or not), so regardless to the outcome of this discussion I would target main branch anyway.

UncleGrumpy commented:
@petermm, I don’t want to disrupt anyone’s workflows. My goal was just to streamline the process for newer users who need to build a custom image to include additional components, like atomvm_mqtt. It makes more sense for me to add an additional cmake job called “deploy” (or similar) that will build the complete image and flash it to the target device using the generated ./build/flasjimage.sh script, leaving the default “build” task unchanged, this will better my goal and should have no affect on any pre-existing workflows.

pguyot commented on CHANGELOG.md:
```diff
@@ -11,6 +11,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### Changed
+
+- ESP32 builds now assemble the complet image automatically if `esp32boot.avm` is found from a previous generic_unix build.
```
s/complet/complete/

pguyot commented:
Should we keep a way to not create the image? (it's not much longer anyway). When working on the VM, I usually only reflash with idf.py flash which is faster than flashing the whole image.

Also, one of my workflows (la machine, obviously) does not include esp32boot.avm at all because it seems to me that esp32boot.avm doesn't bring anything to production images. I'll definitely work around this improvement for newcomers, though.

Likewise, qemu-based tests only use artifacts created by idf.py build (the VM, the partition map and the bootloader) and not the full image.

UncleGrumpy commented:
> Should we keep a way to not create the image?

I converted this back to draft, because I realize after @petermm’s comments that could be intrusive on existing workflows, so instead I will create and document a new build task that can be used as an alternative to `idf.py build` that will create the ready to deploy image, or even flash the image after building so that it flashing a new custom image can be achieved with one command.

pguyot commented:
> > Should we keep a way to not create the image?
> 
> I converted this back to draft, because I realize after @petermm’s comments that could be intrusive on existing workflows, so instead I will create and document a new build task that can be used as an alternative to `idf.py build` that will create the ready to deploy image, or even flash the image after building so that flashing a new custom image can be achieved with one command.

This sounds like a better approach indeed and would be very welcome. I also find it frustrating to have to call the bash script to build the image when I need it.

UncleGrumpy committed:
Allow using the GetVersion.cmake file for any platform

Changes GetVersion.cmake to allow use by platforms other than generic_unix, by allowing the cmake
include of version.cmake before including GetVersion.cmake on other platforms.

Fix version reporting for ESP32 builds when executing mkimage.sh by including the
AtomVM/version.cmake and AtomVM/CMakeModules/GetVersion.cmake files to be able to get the correct
version number when reporting build information during build/mkimage.sh execution.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Streamline ESP32 image creation

Modifies the ESP32 build configuration and mkimage tool to automatically detect the image flavor
based on the partition table used. By checking the offset of the application partition the correct
flavor of language support is used when creating images. Adds a config option to configure an
Elixir supported build by using `idf.py -DATOMVM_ELIXIR_SUPPORT=on set-target ${CHIP}`.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Add esp32boot library to idf.py flash target

Adds the esp32boot library flavor (determined from partition table) to the idf.py flash job,
allowing developers to simply skip running mkimage.sh and flashimage.sh from the build directory
to flash all of the necessary partitions for a complete install to their device.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Fix small bug in ESP32 mkimage.sh when using the help option

Change the script to exit after printing the help, instead of erroneously reporting that the image
was successfully created.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy committed:
Enable optional provisioning of ESP32 nvs partition when building

Adds the option to provision nvs partition data at build time for the esp32 platform. If the file
`nvs_partition.csv` is found in the AtomVM/src/platforms/esp32 directory when building; the nvs
partition will be created and added to the `idf.py flash` task as well as to the configuration for
the mkimage.sh script in the build directory. An example file is provided and `nvs_partition.csv`
is added to `.gitignore` to prevent accidentally pushing secrets to remote repositories.

Signed-off-by: Winford <winford@object.stream>

petermm commented on doc/src/build-instructions.md:
```diff
@@ -272,17 +280,85 @@ $ idf.py build
+ idf.py -p PORT flash
+or
+ python -m esptool --chip esp32 -b 921600 --before default_reset --after hard_reset write_flash --flash_mode dio --flash_size 4MB --flash_freq 40m 0x1000 build/bootloader/bootloader.bin 0x8000 build/partition_table/partition-table.bin 0x10000 build/atomvm-esp32.bin 0x1d0000 ../../../build/libs/esp32boot/esp32boot.avm
+or from the "/home/winford/work/AtomVM/src/platforms/esp32/build" directory
```
from the "/home/winford/work/AtomVM/src/platforms/esp32/build" directory -> from your "AtomVM/src/platforms/esp32/build" directory

or similar..

UncleGrumpy replied:
Thank you!!! That was a last minute addition to the docs (wanted to get the library output in there)… I forgot it had full path info included.

UncleGrumpy replied:
I added some clarification about the path to use, and changed this example output to a less specific path and used the user name "joe", since we started using that as a placeholder for the user name in other documentation examples where a username is expected in program output.

UncleGrumpy commented:
> Should we keep a way to not create the image? (it's not much longer anyway). When working on the VM, I usually only reflash with idf.py flash which is faster than flashing the whole image.

I backed out of assembling the complete image as part of the build process, and it is no longer needed for most users. Instead I added flashing the correct flavor of esp32boot.avm to the `idf.py flash` task. Images can still be created for releases, or if users want to generate a portable image for future use.

In the case of reflashing during development work `idf.py app-flash` may be used to reflash only the AtomVM binary to the `factory` partiton. If flashing an updated bootloader is required, this may be accomplished using `idf.py bootloader-flash` (or combined with app-flash if necessary). This should actually save you a little more time since this does not include the partition map (which is unlikely to be changed often) that is also included when using `idf.py flash`.

petermm commented:
also the simtest/wokwi ci will need some minor adjustment..

petermm commented on UPDATING.md:
```diff
@@ -13,6 +13,20 @@ port socket driver, are also represented by a port and some matching code may ne
 - Ports and pids can be registered. Function `globalcontext_get_registered_process` result now is
 a term that can be a `port()` or a `pid()`.
+- ESP32 builds with Elixir support may be configured without making changes to git tracked files
+using `idf.py -DATOMVM_ELIXIR_SUPPORT=on reconfigure` instead of copying partitions-elixir.csv to
```
I would avoid the implicit esp32 target having been set and go straight for `idf.py -DATOMVM_ELIXIR_SUPPORT=on set-target ${CHIP}` - which works for all scenarios..

This also means a lot of the text can be deleted.

UncleGrumpy commented:
> also the simtest/wokwi ci will need some minor adjustment..

We don’t test Elixir in the simtest, and it has its own sdkconfig.defaults, so I don’t see any chances that need to be made… what did I miss?

petermm commented:
> > also the simtest/wokwi ci will need some minor adjustment..
> 
> We don’t test Elixir in the simtest, and it has its own sdkconfig.defaults, so I don’t see any chances that need to be made… what did I miss?

you are correct, pardon the brainlapse - we probably want tests to mimic builds as much as possible, but let's align in subsequent PR..

UncleGrumpy commented on doc/src/build-instructions.md:
```diff
@@ -270,19 +278,92 @@ You can now build AtomVM using the build command:
-At this point, you can run `idf.py flash` to upload the 3 binaries up to your ESP32 device, and in some development scenarios, this is a preferable shortcut.
+At this point, you can run `idf.py flash` and have a complete working image. In some development
+scenarios is may be helpful to use `idf.py app-flash` or `idf.py bootloader-flash app-flash` to
+avoid re-flashing the entire image if no changes were made to the Erlang or Elixir libraries, and
```
app-flash means flash build/atomvm-{target-chip}.bin, which is the AtomVM executable with libatomvm statically linked in, not the atomvmlib.avm packed BEAM files. To flash those the documentation already describes using `build/flash.sh -l ../../../build/libs/esp32boot/esp32boot.avm`.

UncleGrumpy replied:
To put it another way, it is the contents of the `factory` partition.

UncleGrumpy replied:
I updated this section, hopefully it is more clear now.

UncleGrumpy commented on doc/src/build-instructions.md:
````diff
@@ -239,29 +242,34 @@ $ cd <atomvm-source-tree-root>
 ```shell
-$ idf.py set-target esp32
-$ idf.py reconfigure
+$ idf.py set-target esp32 reconfigure
````
I just used ${CHIP} throughout, but for quoted code examples I think it is fine to use `esp32` as a convention. Especially if ${CHIP} is used in the explanation, so it is obvious that the appropriate chip may be substituted in the code example.

UncleGrumpy commented:
@bettio I don't know what went wrong with the [esp32-build](https://github.com/atomvm/AtomVM/actions/runs/16511473573/job/46694150865), that last push passed on my fork, so I assume it is a fluke.

UncleGrumpy committed:
Enable reproducible builds for ESP32 platform

Use reproducible builds for ESP32 platform so that locally build binaries will be an exact match
of release binaries for the same tag. Error message line info will start from the source path,
trimming any diectories leading up to it. SHA256 sums of binaries built from release tags should be
an exact match of release binaries.

Signed-off-by: Winford <winford@object.stream>

UncleGrumpy commented on UPDATING.md:
```diff
@@ -13,6 +13,17 @@ port socket driver, are also represented by a port and some matching code may ne
+- ESP32 builds with Elixir support may be configured without making changes to git tracked files
+using `idf.py -DATOMVM_ELIXIR_SUPPORT=on set-target ${CHIP}` instead of copying
+partitions-elixir.csv to partitions.csv. This configures the build to use partitions-elixir.csv for
+the partition table. The `boot.avm` offset in the partition table will determine which flavor of
```
Thank you! It should be `main.avm`, I think I was mentally rephrasing this as I was typing and ended up with the wrong partition there.

petermm commented:
This is awesome, and smooths out a LOT of friction..

The `idf.py app-flash` is still confusing to me, the name it self, as well as the description of it. eg. `AtomVM binary` (which doesn't seem to align with docs, which include bootloader, nvs, boot.avm in that term https://github.com/atomvm/AtomVM/blob/21671de0f7b3c87204c932680d65102e7386a664/doc/src/build-instructions.md?plain=1#L432)

I suggest aliasing the app-flash name, and aligning the usage of AtomVM binary, to whatever is deemed correct. Maybe call it platform binary, or factory binary, esp32 factory binary etc. 


```
add_custom_target(platform-flash)
add_dependencies(platform-flash app-flash)
```
(or factory-flash? instead of platform-flash)

thinking of it maybe alias it to factory-flash and call it factory binary

But let's also ship this asap - it's great!

UncleGrumpy commented on doc/src/build-instructions.md:
```diff
@@ -270,19 +275,99 @@ You can now build AtomVM using the build command:
+`factory` partition) to avoid re-flashing the entire image if no changes were made to the Erlang or
+Elixir libraries, and the partition table has not been altered. If you have made changes to the
+sdkconfig file (using `idf.py menuconfig` or by other means) it may be necessary to also update the
+bootloader using `idf.py bootlader-flash`. As with most other `idf.py` commands these may be
```
Thank you, fixed.

UncleGrumpy commented on doc/src/build-instructions.md:
```diff
@@ -270,19 +275,99 @@ You can now build AtomVM using the build command:
+Elixir libraries, and the partition table has not been altered. If you have made changes to the
+sdkconfig file (using `idf.py menuconfig` or by other means) it may be necessary to also update the
+bootloader using `idf.py bootlader-flash`. As with most other `idf.py` commands these may be
+combined (for example: `idf.py bootlader-flash app-flash`). For more information about these
```
fixed

petermm commented on CHANGELOG.md:
```diff
@@ -81,6 +83,11 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- ESP32 Elixir support is determined automatically from the offset of `boot.avm` in the partiton
+table.
+- ESP32 ports now flash a complete working image using the `idf.py flash` task.
+- ESP32 platform now uses reproducable builds.
```
reproducable -> reproducible
